File: src/fit/TimedActionFixture.java
// Modified or written by Object Mentor, Inc. for inclusion with FitNesse.
// Copyright (c) 2002 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.
// Copyright (c) 2002 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.

package fit;

import java.text.SimpleDateFormat;
import java.util.Date;

public class TimedActionFixture extends ActionFixture {
  private static SimpleDateFormat makeDateFormat() {
    //SimpleDateFormat is not thread safe, so we need to create each instance independently.    
    return new SimpleDateFormat("hh:mm:ss");
  }

  public void doTable(Parse table) {
    super.doTable(table);
    table.parts.parts.last().more = td("time");
    table.parts.parts.last().more = td("split");
  }

  public void doCells(Parse cells) {
    Date start = time();
    super.doCells(cells);
    long split = time().getTime() - start.getTime();
    cells.last().more = td(makeDateFormat().format(start));
    cells.last().more = td(split < 1000 ? "&nbsp;" : Double.toString((split) / 1000.0));
  }

  // Utility //////////////////////////////////

  public Date time() {
    return new Date();
  }

  public Parse td(String body) {
    return new Parse("td", gray(body), null, null);
  }

}

File: dotnet/src/fit/TimedActionFixture.cs
// Copyright (C) 2003,2004,2005 by Object Mentor, Inc. All rights reserved.
// Released under the terms of the GNU General Public License version 2 or later.
// Copyright (c) 2002, 2003 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.

using System;

namespace fit 
{
	public class TimedActionFixture : ActionFixture 
	{
		private string format = "hh:mm:ss fff";

		// Traversal ////////////////////////////////

		public override void DoTable(Parse table) 
		{
			base.DoTable(table);
			table.Parts.Parts.Last.More = td("start");
			table.Parts.Parts.Last.More = td("split");
		}

		public override void DoCells(Parse cells)
		{
			DateTime startTime = this.Time();
			base.DoCells(cells);
			DateTime endTime = this.Time();
			long splitMilliseconds = (endTime.Ticks - startTime.Ticks)/10000;
			decimal splitSeconds = new Decimal(splitMilliseconds)/1000;
			cells.Last.More = this.td(startTime.ToString(this.format));
			cells.Last.More = this.td(splitSeconds.ToString("0.000"));
		}

		// Utility //////////////////////////////////

		public virtual DateTime Time()
		{
			return DateTime.Now;
		}

		public virtual Parse td (String body)
		{
			return new Parse("td", Gray(body), null, null);
		}
	}
}